text/0000-public-private-dependencies.md

```diff
@@ -0,0 +1,192 @@
+
+* type leakage through errors. This should be easy to spot for a lint because the
+  wrapper type will implement `std::error::Error`.  The recommendation should most
+  likely be to encourage containing the internal error.
+* traits from other crates.  In particular serde and some other common crates will
+  show up frequently and it might make sense to separately explain types and traits.
+* type leakage through derive.  Users might not be aware they have a dependency to
+  a type when they derive a trait (think `serde_derive`).  The lint might want to
+  call this out separately.
+
+The feature will be called `public_private_dependencies` and it comes with one
+lint flag called `external_private_dependency`.  For all intents and purposes this
+should be the extend of the new terms introduced in the beginning.  This RFC however
+lays the groundwork for later providing aliasing so that a private dependencies could
+be forcefully re-exported as own types.  As such it might make sense to already
+consider what this will be referred to.
+
+It is assumed that this feature will eventually become quite popular due to patterns
+that already exist in the crate ecosystem but it's likely that it will evoke some
+negative opinions initially.  As such it would be a good idea to make a run with
+cargobomb/crater to see what the actual impact of the new linter warnings is and
+how far we are off to making them errors.
+
+crates.io should most likely be updated to render public and private dependencies
+separately.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+I believe that there are no drawbacks if implemented well (this assumes good
+linters and error messages).
+
+# Alternatives
+[alternatives]: #alternatives
+
+For me the biggest alternative to this RFC would be a variation of it where type
+and trait aliasing becomes immediately part of it.  This would meant that a crate
+can have a private dependency and re-export it as its own type, hiding where it
+came from originally.  This would most likely be easier to teach users and can get
+rid of a few "cul-de-sac" situations where users can end up in and their only way
+out is to introduce a public dependency for now.  The assumption is that if trait
+and type aliasing is available the `external_public_dependency` would not need to
+exist.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+There are a few open questions about how to best hook into the compiler and cargo
+infrastructure:
+
+* is passing in the last of public dependencies the correct way to get around it?
+  If yes, what is the parameter supposed to be called.
+* what is the impact of this change going to be. This most likely can be answered
+  running cargobomb/crater.
+* since changing public dependency pins/ranges requires a change in semver it might
+  be worth exploring if cargo could prevent the user in pushing up new crate
+  versions that violate that constraint.
```

Comment:
What if you relax the range on `bar` to include the new version rather than just bumping to only using the latest version. What kind of bump does that mandate?

ISTR seeing that it would be a major change due to cargo always choosing at the high end of the version constraint range. Would it be worth an RFC to cargo to have it see dependencies of `>= 1.1, <= 3` and `^1` in different crates to, instead of deciding on two copies, one of version `2.x` and another of, say, `1.5`, instead see that `1.5` satisfies both and ignore the `2.x` version?